rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,966 @@
+  size_t * count) const
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  if (!count) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  return __get_count(data_readers_, topic_name, count);
+}
+
+static
+std::tuple<std::string, std::string, bool>
+__find_name_and_namespace_from_entity_gid(
+  const GraphCache::ParticipantToNodesMap & participant_map,
+  rmw_gid_t participant_gid,
+  rmw_gid_t entity_gid,
+  bool is_reader)
+{
+  auto it = participant_map.find(participant_gid);
+  if (participant_map.end() == it) {
+    return {"", "", false};
+  }
+  for (const auto & node_info : it->second) {
+    auto & gid_seq = is_reader ? node_info.reader_gid_seq : node_info.writer_gid_seq;
+    auto it = std::find_if(
+      gid_seq.begin(),
+      gid_seq.end(),
+      [&](const rmw_dds_common::msg::Gid & gid) -> bool {
+        return 0u == std::memcmp(gid.data.data(), entity_gid.data, RMW_GID_STORAGE_SIZE);
+      }
+    );
+    if (gid_seq.end() != it) {
+      return {node_info.node_name, node_info.node_namespace, true};
+    }
+  }
+  return {"", "", false};
+}
+
+using DemangleFunctionT = GraphCache::DemangleFunctionT;
+
+static
+rmw_ret_t
+__get_entities_info_by_topic(
+  const GraphCache::EntityGidToInfo & entities,
+  const GraphCache::ParticipantToNodesMap & participant_map,
+  const std::string & topic_name,
+  DemangleFunctionT demangle_type,
+  bool is_reader,
+  rcutils_allocator_t * allocator,
+  rmw_topic_endpoint_info_array_t * endpoints_info)
+{
+  assert(allocator);
+  assert(endpoints_info);
+
+  if (0u == entities.size()) {
+    return RMW_RET_OK;
+  }
+
+  // TODO(ivanpauno): Do something better than overallocating and shrinking to size.
+  rmw_ret_t ret = rmw_topic_endpoint_info_array_init_with_size(
+    endpoints_info,
+    entities.size(),
+    allocator);
+  if (RMW_RET_OK != ret) {
+    return ret;
+  }
+  std::unique_ptr<
+    rmw_topic_endpoint_info_array_t,
+    std::function<void(rmw_topic_endpoint_info_array_t *)>>
+  endpoints_info_delete_on_error(
+    endpoints_info,
+    [allocator](rmw_topic_endpoint_info_array_t * p) {
+      rmw_ret_t ret = rmw_topic_endpoint_info_array_fini(
+        p,
+        allocator
+      );
+      if (RMW_RET_OK != ret) {
+        RCUTILS_LOG_ERROR_NAMED(
+          log_tag,
+          "Failed to destroy endpoints_info when function failed.");
+      }
+    }
+  );
+
+  size_t i = 0;
+  for (const auto & entity_pair : entities) {
+    if (entity_pair.second.topic_name != topic_name) {
+      continue;
+    }
+
+    rmw_topic_endpoint_info_t & endpoint_info =
+      endpoints_info->info_array[i];
+    endpoint_info = rmw_get_zero_initialized_topic_endpoint_info();
+
+    auto result = __find_name_and_namespace_from_entity_gid(
+      participant_map,
+      entity_pair.second.participant_gid,
+      entity_pair.first,
+      is_reader);
+
+    if (!std::get<2>(result)) {
+      // skip endpoints that aren't associated with a node.
+      // TODO(ivanpauno): Do we want to do something else?
```

Comment:
Was it expected to be associated with one, or is like a pure DDS topic that "we" didn't create?

Reply:
That's one case.
The other case is when "we" created it, but a new `ParticipantEntitiesInfo` msg hasn't been received yet. That race condition is a result of combining information of DDS discovery and our custom messages.

There's a way to solve the problem: having a way to identify participants created by "us". I think that we could use "UserData" qos to achieve that (we were using that to communicate the node name and namespace before).

----

The "skip" behavior wasn't nice, I'm doing now the same the current implementation was doing:
https://github.com/ros2/rmw_fastrtps/blob/bd5507b61c7285d6083e514ccd49f00b2945baeb/rmw_fastrtps_shared_cpp/src/rmw_get_topic_endpoint_info.cpp#L193-L206.

Reply:
> There's a way to solve the problem: having a way to identify participants created by "us".

If it helps you detect an error case, then I'd say it is worth doing.

Reply:
I started communicating context name and namespace in https://github.com/ros2/rmw_fastrtps/pull/312/commits/946a9a5d99566a0376842d68f9df615f0d9a58ab and https://github.com/ivanpauno/rmw_dds_common/pull/4/commits/f44c3517c762941c1140c9c5cdb5ffe4937d80f2.

I differentiated both errors in https://github.com/ivanpauno/rmw_dds_common/pull/4/commits/c6dda63c002b73cbff21fa8f5bc1675deccc4365.

Reply:
Maybe, in the case we know it was created by a ros node, and we don't still have the node name, we can just skip it, and it'll appear in a future query.